### PR TITLE
Sort mixin output in order declared

### LINKED ISSUE #2469: Mixins install order isn't fixed which results in cache miss and redundant images
## Describe the bug
When multiple mixins are specified in the definition they are installed in "random" order which doesn't make good use of caching and results in multiple images

## To Reproduce
1. Run `...`
2. Use this porter.yaml from [here](https://gist.github.com/tamirkamara/9389aa07ee22b26bdfa0712b217eb749)
3. Run porter build a few times

## Expected behavior
Expect to see just the first build run every layer and the rest to be cached. But you will probably see a few like the below.

## Porter Command and Output
Notice that `stage-0  5/17` isn't the same in both builds, hence can't be cached. 

Build N
```
 => [internal] load build definition from Dockerfile                                                                                                                     0.0s
 => => transferring dockerfile: 2.12kB                                                                                                                                   0.0s
 => [internal] load .dockerignore                                                                                                                                        0.0s
 => => transferring context: 35B                                                                                                                                         0.0s
 => resolve image config for docker.io/docker/dockerfile-upstream:1.4.0                                                                                                  0.6s
 => CACHED docker-image://docker.io/docker/dockerfile-upstream:1.4.0@sha256:178c4e4a93795b9365dbf6cf10da8fcf517fcb4a17f1943a775c0f548e9fc2ff                             0.0s
 => [internal] load build definition from Dockerfile                                                                                                                     0.0s
 => [internal] load .dockerignore                                                                                                                                        0.0s
 => [internal] load metadata for docker.io/library/debian:buster-slim                                                                                                    0.0s
 => [internal] load build context                                                                                                                                        1.0s
 => => transferring context: 173.47MB                                                                                                                                    0.9s
 => [stage-0  1/17] FROM docker.io/library/debian:buster-slim                                                                                                            0.0s
 => CACHED [stage-0  2/17] RUN useradd nonroot -m -u 65532 -g 0 -o                                                                                                       0.0s
 => CACHED [stage-0  3/17] RUN rm -f /etc/apt/apt.conf.d/docker-clean; echo 'Binary::apt::APT::Keep-Downloaded-Packages "true";' > /etc/apt/apt.conf.d/keep-cache        0.0s
 => [stage-0  4/17] RUN --mount=type=cache,target=/var/cache/apt --mount=type=cache,target=/var/lib/apt     apt-get update &&     apt-get install -y jq="1.5+dfsg-2+b1"  3.0s
 => [stage-0  5/17] RUN --mount=type=cache,target=/var/cache/apt --mount=type=cache,target=/var/lib/apt  apt-get update && apt-get install -y apt-transport-https lsb-  15.7s
 => [stage-0  6/17] RUN curl -sL https://packages.microsoft.com/keys/microsoft.asc | gpg --dearmor > /etc/apt/trusted.gpg.d/microsoft.asc.gpg                            0.9s
 => [stage-0  7/17] RUN echo "deb [arch=amd64] https://packages.microsoft.com/repos/azure-cli/ $(lsb_release -cs) main" > /etc/apt/sources.list.d/azure-cli.list         1.1s
 => [stage-0  8/17] RUN --mount=type=cache,target=/var/cache/apt --mount=type=cache,target=/var/lib/apt  apt-get update && apt-get install -y --no-install-recommends   22.3s
 => [stage-0  9/17] RUN apt-get update && apt-get install -y wget unzip &&  apt-get clean -y && rm -rf /var/lib/apt/lists/* &&  wget https://releases.hashicorp.com/ter  7.6s
 => [stage-0 10/17] COPY terraform/ /cnab/app/terraform/                                                                                                                 0.1s
 => [stage-0 11/17] RUN cd /cnab/app/terraform &&  terraform init -backend=false &&  rm -fr .terraform/providers &&  terraform providers mirror /usr/local/share/terra  27.3s
 => [stage-0 12/17] COPY --link . /cnab/app                                                                                                                              0.3s
 => [stage-0 13/17] RUN rm /cnab/app/porter.yaml                                                                                                                         0.6s
 => [stage-0 14/17] RUN rm -fr /cnab/app/.cnab                                                                                                                           0.6s
 => [stage-0 15/17] COPY --link .cnab /cnab                                                                                                                              0.2s
 => [stage-0 16/17] RUN chgrp -R 0 /cnab && chmod -R g=u /cnab                                                                                                           1.1s
 => [stage-0 17/17] WORKDIR /cnab/app                                                                                                                                    0.1s
 => exporting to image                                                                                                                                                   6.9s
 => => exporting layers                                                                                                                                                  6.9s
 => => writing image sha256:cc4b2d7987b6531fca5bc1b6abf0ca45e27a19527d0b9383da8112e79e7c6c14                                                                             0.0s
 => => naming to docker.io/azuretre/tre-workspace-base:porter-785f646f84ea65af983a9191951a9a24 
```

Build N+1
```
  => [internal] load build definition from Dockerfile                                                                                                                     0.0s
 => => transferring dockerfile: 2.12kB                                                                                                                                   0.0s
 => [internal] load .dockerignore                                                                                                                                        0.0s
 => => transferring context: 35B                                                                                                                                         0.0s
 => resolve image config for docker.io/docker/dockerfile-upstream:1.4.0                                                                                                  0.6s
 => CACHED docker-image://docker.io/docker/dockerfile-upstream:1.4.0@sha256:178c4e4a93795b9365dbf6cf10da8fcf517fcb4a17f1943a775c0f548e9fc2ff                             0.0s
 => [internal] load .dockerignore                                                                                                                                        0.0s
 => [internal] load build definition from Dockerfile                                                                                                                     0.0s
 => [internal] load metadata for docker.io/library/debian:buster-slim                                                                                                    0.0s
 => [internal] load build context                                                                                                                                        1.0s
 => => transferring context: 173.47MB                                                                                                                                    1.0s
 => [stage-0  1/17] FROM docker.io/library/debian:buster-slim                                                                                                            0.0s
 => CACHED [stage-0  2/17] RUN useradd nonroot -m -u 65532 -g 0 -o                                                                                                       0.0s
 => CACHED [stage-0  3/17] RUN rm -f /etc/apt/apt.conf.d/docker-clean; echo 'Binary::apt::APT::Keep-Downloaded-Packages "true";' > /etc/apt/apt.conf.d/keep-cache        0.0s
 => CACHED [stage-0  4/17] RUN --mount=type=cache,target=/var/cache/apt --mount=type=cache,target=/var/lib/apt     apt-get update &&     apt-get install -y jq="1.5+dfs  0.0s
 => [stage-0  5/17] RUN apt-get update && apt-get install -y wget unzip &&  apt-get clean -y && rm -rf /var/lib/apt/lists/* &&  wget https://releases.hashicorp.com/ter  9.3s
 => [stage-0  6/17] COPY terraform/ /cnab/app/terraform/                                                                                                                 0.1s
 => [stage-0  7/17] RUN cd /cnab/app/terraform &&  terraform init -backend=false &&  rm -fr .terraform/providers &&  terraform providers mirror /usr/local/share/terra  31.3s
 => [stage-0  8/17] RUN --mount=type=cache,target=/var/cache/apt --mount=type=cache,target=/var/lib/apt  apt-get update && apt-get install -y apt-transport-https lsb-r  9.7s
 => [stage-0  9/17] RUN curl -sL https://packages.microsoft.com/keys/microsoft.asc | gpg --dearmor > /etc/apt/trusted.gpg.d/microsoft.asc.gpg                            1.0s
 => [stage-0 10/17] RUN echo "deb [arch=amd64] https://packages.microsoft.com/repos/azure-cli/ $(lsb_release -cs) main" > /etc/apt/sources.list.d/azure-cli.list         0.5s
 => [stage-0 11/17] RUN --mount=type=cache,target=/var/cache/apt --mount=type=cache,target=/var/lib/apt  apt-get update && apt-get install -y --no-install-recommends   14.1s
 => [stage-0 12/17] COPY --link . /cnab/app                                                                                                                              0.3s
 => [stage-0 13/17] RUN rm /cnab/app/porter.yaml                                                                                                                         0.5s
 => [stage-0 14/17] RUN rm -fr /cnab/app/.cnab                                                                                                                           0.5s
 => [stage-0 15/17] COPY --link .cnab /cnab                                                                                                                              0.2s
 => [stage-0 16/17] RUN chgrp -R 0 /cnab && chmod -R g=u /cnab                                                                                                           0.9s
 => [stage-0 17/17] WORKDIR /cnab/app                                                                                                                                    0.1s
 => exporting to image                                                                                                                                                   8.5s
 => => exporting layers                                                                                                                                                  8.4s
 => => writing image sha256:26b768ddab37d803ceb0313a195f1361690870362551f5d2b92958d99e040f7e                                                                             0.0s
 => => naming to docker.io/azuretre/tre-workspace-base:porter-785f646f84ea65af983a9191951a9a24 
```

## Version
Copy the output of `porter version` below

[porter v1.0.2 (0656f1fd)]

### PR DESCRIPTION
**Pull request by carolynvs**
# What does this change
When we are querying mixins, such as when we are asking each mixin for Dockerfile lines to include in the bundle image, we call them in go routines. Depending on how quickly they return, the order is not guaranteed and may change. I have updated our mixin query function to sort the results of querying the mixins in a consistent order, the order that they are defined in the bundle.

During build in particular, this ensures that we optimize layer caching as much as possible by always outputing the same contents of the Dockerfile when building a bundle. Previously, it could include the mixins in varying order which prevented reuse of layers from previous builds.

# What issue does it fix
Fixes #2469

# Notes for the reviewer
N/A

# Checklist
- [x] Did you write tests?
- [ ] Did you write documentation?
- [ ] Did you change porter.yaml or a storage document record? Update the corresponding schema file.
- [ ] If this is your first pull request, please add your name to the bottom of our [Contributors][contributors] list. Thank you for making Porter better! 🙇‍♀️

# Reviewer Checklist
* Comment with /azp run test-porter-release if a magefile or build script was modified
* Comment with /azp run porter-integration if it's a non-trivial PR

[contributors]: https://getporter.org/src/CONTRIBUTORS.md

### CODE REVIEW

File: pkg/mixin/query/query.go [hunk 1]
```diff
@@ -46,39 +46,51 @@ type MixinInputGenerator interface {
...
 	gerr := errgroup.Group{}
 
-	for _, mn := range mixinNames {
-		mixinName := mn // Force mixinName to be in the go routine's closure below
+	for i, mn := range mixinNames {
+		// Force variables to be in the go routine's closure below
+		i := i
+		mn := mn
+		results[i].Name = mn
+		resultLookup[mn] = i
```
Comment (sgettys) [inline pkg/mixin/query/query.go:79]: This looks unused, it is here for future capabilities to lookup the index of a mixin name?

**Review by sgettys — APPROVED** [review]

File: pkg/mixin/query/query.go (hunk 1 above)
Comment (carolynvs) [reply pkg/mixin/query/query.go:79]: Oops! I thought I needed it but then after making some more changes, you're right it's not used anymore. I'll remove it.

**Review by sgettys — APPROVED** [review]

**Comment by carolynvs** [comment]
I just realized that the way I'm using the map isn't safe to use with go routines. I'll revert some of it to go back to using channels.

**Comment by carolynvs** [comment]
Happily, the problem was just in the unit tests, not the code. I've fixed the test helper to avoid the race.

File: pkg/mixin/helpers.go [hunk 2]
```diff
@@ -53,19 +53,19 @@ func NewTestMixinProvider() *TestMixinProvider {
...
 
-func (p *TestMixinProvider) PrintExecOutput(pkgContext *portercontext.Context, name string, commandOpts pkgmgmt.CommandOptions) error {
+func (p *TestMixinProvider) PrintMixinOutput(pkgContext *portercontext.Context, name string, commandOpts pkgmgmt.CommandOptions) error {
 	switch commandOpts.Command {
 	case "build":
 		if p.ReturnBuildError {
 			return errors.New("encountered build error")
 		}
-		fmt.Fprintln(pkgContext.Out, "# exec mixin has no buildtime dependencies")
+		fmt.Fprintf(pkgContext.Out, "# %s mixin has no buildtime dependencies", name)
```
Comment (VinozzZ) [inline pkg/mixin/helpers.go:68]: should we preserve the new line break in the output here?

File: pkg/pkgmgmt/client/helpers.go [hunk 3]
```diff
@@ -20,20 +21,23 @@ type TestPackageManager struct {
...
-	called map[string]int
+	called sync.Map
+	lock   sync.Mutex
 }
 
 // GetCalled tracks how many times each package was called
-func (p *TestPackageManager) GetCalled() map[string]int {
-	return p.called
+func (p *TestPackageManager) GetCalled(mixin string) int {
+	calls, _ := p.called.Load(mixin)
```
Comment (VinozzZ) [inline pkg/pkgmgmt/client/helpers.go:30]: should we check the boolean returned by Load to make sure the mixin exist in the map before we do the type assertion?

**Comment by carolynvs** [comment]
@VinozzZ I've updated the PR with your feedback, please take another look

**Review by VinozzZ — APPROVED** [review]
lgtm!

**Pull request merged**